HWKALERTS-207 Allow ExternalCondition to be fired on Event submission

lucasponce opened the pull request:
Extending the ExternalCondition to events.
Please @jshaughn, can you review it ?
Thanks.

jshaughn commented on hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/model/condition/ExternalConditionEval.java:
```diff
@@ -37,6 +40,10 @@
 
+    private enum TypeValue {
+        DATA, EVENT
+    }
```
Do we really need the new enum or to store the Event in a field?  It may be sufficient to just add the new constructor for Event?  You have a String value using Data or Event.

If you really want to keep it, perhaps change from 'TypeValue' to 'ValueType'.

lucasponce replied:
I'm not sure. I'm +1 to change to ValueType.
As you suggest, we can use the presence of the String/Event as implicit discriminator about it.
I'm not against that either, just wondering in some corner cases (if any) what happen if there is a null value on both (then perhaps is an IllegalArgumentException).
I'm for maintain temporarily this enum changing to ValueType and we can clean it later once the events alerter is putting in place (we can review it).

jshaughn commented on hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/model/condition/ExternalConditionEval.java:
```diff
@@ -18,12 +18,15 @@
+
 /**
  * An evaluation state for an external condition.  Note that external conditions report values as
  * <code>StringData</code>.
```
You can remove "  Note that external conditions report values as <code>StringData</code>."   That is an obsolete comment.

lucasponce replied:
+1

jshaughn commented on hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/model/condition/ExternalConditionEval.java:
```diff
@@ -47,16 +54,33 @@
+        super(Type.EXTERNAL, condition.match(event.getText()), event.getCtime(), event.getContext());
+        this.condition = condition;
+        this.event = event;
+        this.typeValue = TypeValue.DATA;
```
If you keep the typeValue field I think this should be EVENT.  But as I said above, I don't think you need it.

jshaughn commented on hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/model/condition/ExternalConditionEval.java:
```diff
@@ -47,16 +54,33 @@
         super(Type.EXTERNAL, condition.match(data.getValue()), data.getTimestamp(), data.getContext());
         this.condition = condition;
         this.value = data.getValue();
+        this.typeValue = TypeValue.EVENT;
```
And if you keep this field, I think this one should be DATA.

jshaughn commented on hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/model/condition/ExternalConditionEval.java:
```diff
@@ -75,6 +99,14 @@ public void setValue(String value) {
+
+    public void setEvent(Event event) {
+        this.event = event;
+    }
```
If you keep the 'event'  and 'valueType' fields, this setter should probably set the valueType to EVENT for safe deserialization.

jshaughn commented on hawkular-alerts-api/src/main/java/org/hawkular/alerts/api/model/condition/ExternalConditionEval.java:
```diff
@@ -97,42 +129,48 @@ public int getConditionSetIndex() {
     @Override
     public String getLog() {
-        return condition.getLog(value) + ", evalTimestamp=" + evalTimestamp + ", dataTimestamp=" + dataTimestamp;
+        switch (typeValue) {
```
If you keep the event you could just use its existence here as a discriminator.

lucasponce commented:
Finally I removed the enum.
I think we have room to introduce it later if there is some use case where is needed.
Added a missing json deserializer task for new event field on ExternalConditionEval.
